### DISCUSSION TITLE
Getting the result of Data Annotations validation

### DISCUSSION THREAD

**Discussion by JDr111** [AvaloniaUI/Avalonia#17373] [Q&A]
Suppose there's a simple user input form, with a confirmation button. The form's model is validated using Data Annotation attributes (System.ComponentModel.DataAnnotations) and Avalonia's respective built-in validation plugin. How does one obtain the result of the validation process (a boolean value is enough) in order to enable or disable the confirmation button, preventing the user from continuing if there's validation errors?

**Comment by ProviceUnify** [comment]
did something work out?

**Reply by ProviceUnify** [reply]
In communitytoolkit.mvvm you can use `GetErrors("PropName")` inside ViewModel. It should return *List* of validation errors
![image](https://github.com/user-attachments/assets/e2237de8-920e-4dcc-948b-7560a00fb255)

May be System.ComponentModel.DataAnnotations has something like this?

My full *working* realisation below:
```
Avalonia: 11.2.1
.NET: 9 (available in 8 also i think)
CommunityToolkit.MVVM
```
```xml
<!-- MyView.axaml -->
<TextBox Text="{Binding NewPassword}" />
```

```csharp
// MyViewModel.cs
public partial class MyViewModel : ViewModelBase // -> : ObservableValidator <-
{
    [ObservableProperty]
    [NotifyDataErrorInfo]
    [RegularExpression(passwordRegEx, ErrorMessage = "Blah-blah")]
    private string _newPassword;

    private void AnyMethod() {
         var list = /*this.*/GetErrors("NewPassword");
    }
}
```

**Reply by JDr111** [reply]
Essentially the same thing as your working realization, just a bit more verbose. I ended up deriving all my model classes from the "[ObservableValidator](https://learn.microsoft.com/en-us/dotnet/api/microsoft.toolkit.mvvm.componentmodel.observablevalidator?view=win-comm-toolkit-dotnet-7.1)" class in the Community MVVM Toolkit, which does exactly what I need.

Looking at the MS documentation, ObservableValidator implements the INotifyDataErrorInfo interface, which exposes the "HasErrors" property and "GetErrors" and "ClearErrors" methods, which provide all the functionality I needed. It also exposes the "Validate" method for running validation manually, for instance to check for validation errors of a configuration form before saving data to a database.

Whatever fancier validation logic I needed I achieved by extending ValidationAttribute base class, and then implementing custom validation attributes deriving from the extended ValidationAttribute base class. For instance: "ValidationAttribute (base class in .NET) -> ValidationExtendedAttribute -> CustomUrlAttribute". The validator implemented by ObservableValidator works fine with a setup like this.

This way you're also able to utilize localization capabilities of the validation attributes by specifying resource type and resource names in the attribute parameters (and really should be _the_ way to display validation errors in UI, localized or not):

```c#
private string authorizeUrl;

[Required(AllowEmptyStrings = false, ErrorMessageResourceType = typeof(Localization.Validation), ErrorMessageResourceName = nameof(Localization.Validation.FieldCannotBeEmpty))]
[CustomUrl(ErrorMessageResourceType = typeof(Localization.Validation), ErrorMessageResourceName = nameof(Localization.Validation.UrlInvalid))]
public string AuthorizeUrl
{
    get => authorizeUrl;
    set => SetProperty(ref authorizeUrl, value, true);
}
``` 
("Localization.Validation" is a resource class generated by the RESX source generator.)

Implementation of the "CustomUrlAttribute" (the only difference from UrlAttribute in .NET is that the custom one uses Regex source generation to ensure compatibility with NativeAOT; I also plan on adding further changes to the Regex itself once I get there, because this one doesn't accept an IP address in place of a FQDN):
```c#
[AttributeUsage(AttributeTargets.Property | AttributeTargets.Field | AttributeTargets.Parameter, AllowMultiple = false)]
public partial class CustomUrlAttribute : ValidationExtendedAttribute
{
    private const string pattern = @"^(https?|ftp):\/\/(((([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:)*@)?(((\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5])\.(\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5])\.(\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5])\.(\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5]))|((([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))\.)+(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))\.?)(:\d*)?)(\/((([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:|@)+(\/(([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:|@)*)*)?)?(\?((([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:|@)|[\uE000-\uF8FF]|\/|\?)*)?(\#((([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(%[\da-f]{2})|[!\$&'\(\)\*\+,;=]|:|@)|\/|\?)*)?$";

    [GeneratedRegex(pattern, RegexOptions.IgnoreCase | RegexOptions.ExplicitCapture)]
    private static partial Regex GetUrlRegex();

    public override bool IsValid(object value)
    {
        if (value == null)
        {
            return true;
        }

        string valueAsString = value as string;

        if (valueAsString != null)
        {
            Regex regex = GetUrlRegex();
            return valueAsString != null && regex.Match(valueAsString).Length > 0;
        }
        else
        {
            return false;
        }
    }
}
``` 

Then it's just a matter of binding to Boolean properties of the ViewModel which check whether all models have or don't have any errors, like @trashhacker suggested, and monitoring the "ErrorsChanged" event of the ObservableValidator models, within the ViewModel:
```c#
public bool ButtonEnabled => model1.HasErrors == false && model2.HasErrors == false ...;
...
private void ErrorsChanged(object sender, DataErrorsChangedEventArgs e)
{
    OnPropertyChanged(nameof(ButtonEnabled));
}
```

Then just bind in XAML:
```html
<UserControl ...
    xmlns:loc="using:My.Project.Localization"
    xmlns:vm="using:My.Project.ViewModels"
    x:DataType="vm:ConfigurationViewModel"
...
<Button IsEnabled={Binding ButtonEnabled}>
    <TextBlock Text={x:Static loc:Configuration:Save}/>
</Button>
``` 

I like this approach, it makes all the pieces play along nicely.

**Comment by trashhacker** [comment]
u have to implement the validation of the annotaged property. 
something like Notify(ref propertyfield, value, true); 
whe your notify method call the Validate method of the Attribute and store the result

maybe like that


        public static bool Validate(PropertyInfo property, ValidationAttribute attribute, object? value, out string? message)
        {
            if (attribute.IsValid(value))
            {
                message = null;
                return false;
            }

            string displayname = SelectName(property);

            if (string.IsNullOrWhiteSpace(attribute.ErrorMessageResourceName) == false)
            {
                if (resourcemanager.GetString(attribute.ErrorMessageResourceName, CultureInfo.CurrentCulture) is string stringformat)
                {
                    message = FormatString(attribute, displayname, stringformat);
                    return true;
                }
            }

            if (string.IsNullOrWhiteSpace(attribute.ErrorMessage) == false)
            {
                message = attribute.ErrorMessage;
                return true;
            }

            message = $"\"{value}\" is not valid for {displayname}";
            return true;
        }




the u implement something like or INotifyDataErrorInfo and its functionallity
then u could do something like 

publick bool ButtonErrors => GetErrors("PropertyName");
and do the ReactiveUI thing if u use it
